feat: `soar` module

fiftydinar opened the pull request:
Closes: https://github.com/blue-build/modules/issues/412

I tested everything, including auto-upgrade of packages & it works well.

@azathothas For any suggestions or fixes, please write

fiftydinar committed:
feat: `soar` module

fiftydinar committed:
fix(soar): Inverted `unlock-repos` logic

fiftydinar committed:
chore(soar): Add `soar` to modules.json

gmpinder commented on modules/soar/README.md:
```diff
@@ -0,0 +1,110 @@
+  - `sudo cp /usr/share/systemd/system/soar-upgrade-packages.service /etc/systemd/system/soar-upgrade-packages.service`
+- by copying the timer file:
+  - `sudo cp /usr/share/systemd/system/soar-upgrade-packages.timer /etc/systemd/system/soar-upgrade-packages.timer`
+- by modifying the systemd service in `/etc/` to contain the custom path to the config file in `Exec`
```
You're copying the files to `/usr/lib/systemd/system` in the script.

fiftydinar committed:
fix(soar): Logic for detecting if `unlock-repos` is disabled

fiftydinar committed:
fix(docs): Wrong `/usr/` systemd path reference

fiftydinar committed:
chore(soar): Add log for `unlock_repos`

fiftydinar committed:
chore(soar): Add `finished=false` at the beggining of `soar` shell profile

fiftydinar committed:
fix(soar): `/root` being `$HOME` in config

fiftydinar committed:
fix(soar): Wrong `PATH`

I forgot to add `bin` directory

fiftydinar committed:
docs(soar): Clarify that `PATH` needs to be exported only outside of standard config dir

Azathothas commented on modules/soar/soar.sh:
```diff
@@ -0,0 +1,79 @@
+fi
+
+# Configuration for unlocking all default repos (outside of 'bincache')
+UNLOCK_REPOS=$(echo "${1}" | jq -r 'try .["unlock-repos"]')
```
Other than the default repos, 
there's also a massive collection of external repos, which can be enabled with
`soar defconfig --external` (only if soar defconfig wasn't already ran before)

Azathothas commented:
I also created an issue to configure some of these things you are doing with sed, via environment variable.
https://github.com/pkgforge/soar/issues/54

fiftydinar committed:
chore(soar): Unlock extra repos if `unlock-repos` is enabled

fiftydinar committed:
docs(soar): Mention of external repos

fiftydinar committed:
chore(soar): Simplify getting `XDG` directories

fiftydinar committed:
chore(soar): Add removing config in uninstallation script

fiftydinar committed:
chore(soar): Remove whole config directory instead of file in uninstallation script

fiftydinar committed:
fix(soar): Missing quote in uninstallation script

fiftydinar committed:
chore(soar): Make uninstallation script POSIX

fiftydinar committed:
fix(soar): Bad syntax for `soar-upgrade-packages` service `exec`

It's `ExecStart`, not `Exec`.

fiftydinar committed:
fix(soar): Set `auto-upgrade` timer as `--user` instead of `--system`

fiftydinar committed:
chore(soar): Shorten the boot delay & randomize it for `auto-upgrades`

fiftydinar committed:
chore(soar): Add `network-online.target` in `Wants` to systemd service for `auto-upgrades`

fiftydinar committed:
chore(soar): Remove `upgrade-wait-after-boot`, as good defaults are set with no need for changing

fiftydinar committed:
chore(soar): Add option for unlocking the `PATH` in all conditions

fiftydinar committed:
chore(soar): Add `unlock-path` option to typespec

fiftydinar committed:
docs(soar): Add `unlock-path` to `module.yml` example

fiftydinar committed:
chore(soar): Remove condition for checking the `soar.sh` file

If another `soar` module overwrites the previous one, it should be able to overwrite `soar.sh` shell profile too with custom option if set.

fiftydinar committed:
chore(soar): Remove `unlock-path`, as it's too dangerous

fiftydinar committed:
chore(soar): Don't copy update timer & service if `auto-upgrade` is false

fiftydinar committed:
chore(soar): Style fix

fiftydinar committed:
Merge branch 'main' into soar-module

fiftydinar committed:
Merge branch 'main' into soar-module

fiftydinar committed:
chore(soar): Don't execute `upgrade` when soar is already running

fiftydinar committed:
chore(soar): Use environment variable for `SOAR_CONFIG` to simplify things

fiftydinar committed:
docs(sour): Fix typo

QaidVoid commented on modules/soar/soar.sh:
```diff
@@ -0,0 +1,65 @@
+  mkdir -p "/usr/share/bluebuild/soar"
+  soar defconfig -c "/usr/share/bluebuild/soar/config.toml"
+  # Remove all other repositories except bincache
+  sed -i '/^\[\[repositories\]\]/{:a;N;/name = "bincache"/!{/\n[[:space:]]*$/!ba;d}}' "/usr/share/bluebuild/soar/config.toml"
```
Can be reduced to 
```sh
soar -c /usr/share/bluebuild/soar/config.toml defconfig -r bincache
```

fiftydinar committed:
chore(soar): Use native `soar` command for setting `bincache` repo only

fiftydinar committed:
fix(soar): Make `soar` condition check for  auto-upgrades more reliable

fiftydinar committed:
chore(soar): Some more improvements to auto-update condition check

False positive can happen if some text editor edits some file like `soar-something.txt` or similar.

This assures more that only executed binary is detected.

xynydev commented on modules/soar/soar.sh:
```diff
@@ -0,0 +1,63 @@
+  soar -c /usr/share/bluebuild/soar/config.toml defconfig -r bincache
+  # Fix /root being ${HOME}
+  sed -i 's|/root|~|g' "/usr/share/bluebuild/soar/config.toml"
+fi
```
The `mkdir` and the `sed` commands could be lifted out of the `if` clause to prevent repetition.

xynydev commented on modules/soar/soar-upgrade-packages.service:
```diff
@@ -0,0 +1,9 @@
+[Service]
+Type=oneshot
+ExecCondition=/bin/bash -c 'if ps aux | grep -v grep | grep -E -q " /sbin/soar | /bin/soar | /usr/sbin/soar | /usr/bin/soar | soar "; then exit 1; else exit 0; fi'
+ExecStart=/bin/soar update
```
Is it always at `/bin/soar`? Or maybe this is something `systemd` allows you to do? Looking at the installation script, seems like it should be `/usr/bin/soar`...

fiftydinar replied:
`/bin` is a symlink to `/usr/bin`. Same case for `/sbin` & `/usr/sbin` being symlinks to those 2. So that's all the same thing.

xynydev commented:
Overall looking pretty good. Soar seems like an interesting project, had not heard of it before. Documentation still needs some structural and spelling fixes, plus a link to https://pkgs.pkgforge.dev/?repo=soarpkgs. I'll work on that before merge, hopefully during next week at least. 

Am I correct in that `soar` is mainly for installing packages separately for each user into `~/.local/bin/` or something like that? Has the decision to not offer automatic installation of a specific package set on an image been justified and thought through?

fiftydinar commented:
> Overall looking pretty good. Soar seems like an interesting project, had not heard of it before. Documentation still needs some structural and spelling fixes, plus a link to https://pkgs.pkgforge.dev/?repo=soarpkgs. I'll work on that before merge, hopefully during next week at least.

:+1: 

> Am I correct in that `soar` is mainly for installing packages separately for each user into `~/.local/bin/` or something like that?

Correct, it installs the binaries in `${XDG_DATA_HOME}/soar/bin/` by default.

> Has the decision to not offer automatic installation of a specific package set on an image been justified and thought through?

It's mostly the same reason as for `brew` & that's the potential complexity in implementing it, which would delay merging this module.
But looking at it, I can say that it would be easier to implement that here than it would be for `brew` module.
That can come sometime later after this PR if needed imo.

fiftydinar committed:
chore(soar): Remove duplicate code for setting config

gmpinder commented on modules/soar/module.yml:
```diff
@@ -0,0 +1,5 @@
+shortdesc: The soar module installs & integrates soar package manager, as an alternative to Homebrew / Linuxbrew.
+example: |
+  type: soar
+  unlock-repos: true
```
Could you add a bigger example here?

fiftydinar replied:
Added

fiftydinar committed:
docs(soar): Add full example of options in `module.yml`

fiftydinar committed:
Merge branch 'main' into soar-module

fiftydinar commented:
**TODO**:

- Docs - external repos:
  - Add a reference & note to the pkgforge page about external repos, which lists which external repos are available & their trust + security status.
  https://docs.pkgforge.dev/repositories/external
  
- `unlock-repos`:
  - Rename `unlock-repos` to `additional-repos` or `external-repos`, because wording is a bit misleading.
`additional-repos` is a bit more accurate, because there are also official repos like `pkgcache`, which are not external & are not enabled by default.
Additionally, we are not locking out anything, just disabling all those default & external repos in favor of `bincache` repo for CLI binaries distribution.
  - However, besides `bincache`, there are also very useful external repos like `pkgforge-cargo` & `pkgforge-go` for Rust & Go static binaries, which is a nice collection. Even `bluebuild` CLI can be installed from there :D Maybe worth considering to enable those too?
  
There are also some CLI binaries in `pkgcache` as an AppImage, like `file` or `git` program, which had to be built that way, but those are in very small amount, so I would not enable `pkgcache` repo just for that.

I won't do it in close time interval, as I'm on a pause from tech for some time as I said on Discord.
I just had to use tech today to fix some stuff, so I figured to also do this before I finally leave it.

fiftydinar committed:
Merge branch 'main' into soar-module

fiftydinar committed:
chore(soar): Rename `unlock-repos` to `additional-repos` & add docs about repo info

fiftydinar committed:
chore(soar): Enable `pkforge-cargo` & `pkgforge-go` repos

fiftydinar commented:
> **TODO**:
> 
> * Docs - external repos:
>   
>   * Add a reference & note to the pkgforge page about external repos, which lists which external repos are available & their trust + security status.
>     https://docs.pkgforge.dev/repositories/external
> * `unlock-repos`:
>   
>   * Rename `unlock-repos` to `additional-repos` or `external-repos`, because wording is a bit misleading.
>     `additional-repos` is a bit more accurate, because there are also official repos like `pkgcache`, which are not external & are not enabled by default.
>     Additionally, we are not locking out anything, just disabling all those default & external repos in favor of `bincache` repo for CLI binaries distribution.
>   * However, besides `bincache`, there are also very useful external repos like `pkgforge-cargo` & `pkgforge-go` for Rust & Go static binaries, which is a nice collection. Even `bluebuild` CLI can be installed from there :D Maybe worth considering to enable those too?
> 
> There are also some CLI binaries in `pkgcache` as an AppImage, like `file` or `git` program, which had to be built that way, but those are in very small amount, so I would not enable `pkgcache` repo just for that.
> 
> I won't do it in close time interval, as I'm on a pause from tech for some time as I said on Discord. I just had to use tech today to fix some stuff, so I figured to also do this before I finally leave it.

I did all of the above, including enabling `pkgforge-cargo` & `pkgforge-go` repos.

fiftydinar committed:
Merge branch 'main' into soar-module

xynydev committed:
docs: rewordings, run through languagetool

xynydev commented on modules/soar/README.md:
```diff
@@ -0,0 +1,64 @@
+Other default & external repos that contain AppImages & other similar formats are disabled to make `soar` focused on CLI binaries only.
+This is configurable if you wish to have a package manager for GUI applications, see [`Configuration options`](#configuration-options).
+
+Those binaries are built from the GitHub's cloud registry.
```
What are "those binaries"? Are they built from "GitHub's cloud registry" or are they published to "GitHub's Container registry" (ghcr.io)?

fiftydinar replied:
Yes, published is a better word.

"Those binaries" includes all available static binaries from pkgforge

xynydev replied:
Cleared up in [2ba806f](https://github.com/blue-build/modules/pull/413/commits/2ba806f1deaa4c03e03dadb460bc1811187d128e)

fiftydinar committed:
chore(soar): Add condition in shell profile to check if `soar` PATH already exists

xynydev committed:
docs: final reword

fiftydinar committed:
chore(soar): Disable redundant shellcheck SC2076 in shell profile

fiftydinar committed:
Merge branch 'main' into soar-module

xynydev committed:
Merge branch 'main' into soar-module
